modules/host/contractmanager/sectoradd.go

```diff
@@ -0,0 +1,216 @@
+package contractmanager
+
+import (
+	"encoding/binary"
+	"math"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/crypto"
+	"github.com/NebulousLabs/Sia/modules"
+)
+
+// mostSignificantBit returns the index of the most significant bit of an input
+// value.
+func mostSignificantBit(i uint64) uint64 {
+	if i == 0 {
+		panic("no bits set in input")
+	}
+
+	bval := []uint64{0, 0, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7}
+	r := uint64(0)
+	if i&0xffffffff00000000 != 0 {
+		r += 32
+		i = i >> 32
+	}
+	if i&0x00000000ffff0000 != 0 {
+		r += 16
+		i = i >> 16
+	}
+	if i&0x000000000000ff00 != 0 {
+		r += 8
+		i = i >> 8
+	}
+	if i&0x00000000000000f0 != 0 {
+		r += 4
+		i = i >> 4
+	}
+	return r + bval[i]
+}
+
+// randFreeSector will take a usage array and find a random free sector within
+// the usage array. If one is found, the bit will be flipped atomically in the
+// usage array to indicate that the sector is no longer available. The uint64
+// indicates the index of the sector within the usage array.
+func randFreeSector(usage []uint64) uint32 {
+	// Pick a random starting location. Scanning the sector in a short amount
+	// of time requires starting from a random place.
+	start, err := crypto.RandIntn(len(usage))
+	if err != nil {
+		panic(err)
+	}
+
+	// Find the first element of the array that is not completely full.
+	var i int
+	for i = start; i < len(usage); i++ {
+		if usage[i] != math.MaxUint64 {
+			break
+		}
+	}
+	// If nothing was found by the end of the array, a wraparound is needed.
+	if i == len(usage) {
+		for i = 0; i < start; i++ {
+			if usage[i] != math.MaxUint64 {
+				break
+			}
+		}
+	}
```

Comment:
these loops could be combined using a modulo, but that might incur unacceptable overhead